library/eddsa.c

```diff
@@ -0,0 +1,446 @@
+                          int (*f_rng)(void *, unsigned char *, size_t), void *p_rng,
+                          mbedtls_eddsa_id eddsa_id,
+                          const unsigned char *ed_ctx, size_t ed_ctx_len)
+{
+    int ret = 0;
+    mbedtls_mpi h;
+    mbedtls_ecp_point sB, hA, R;
+
+    mbedtls_mpi_init( &h );
+    mbedtls_ecp_point_init( &sB );
+    mbedtls_ecp_point_init( &hA );
+    mbedtls_ecp_point_init( &R );
+
+    if( mbedtls_mpi_cmp_mpi( s, &grp->N ) >= 0 || mbedtls_mpi_cmp_int( s, 0 ) < 0 )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+#ifdef MBEDTLS_ECP_DP_ED25519_ENABLED
+    if( grp->id == MBEDTLS_ECP_DP_ED25519 )
+    {
+        mbedtls_sha512_context sha_ctx;
+        unsigned char sha_buf[64], tmp_buf[32];
+        size_t olen = 0;
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary_le( r, tmp_buf, sizeof(tmp_buf) ) );
+        MBEDTLS_MPI_CHK( mbedtls_ecp_point_read_binary( grp, &R, tmp_buf, sizeof(tmp_buf) ) );
+
+        mbedtls_sha512_init( &sha_ctx );
+        mbedtls_sha512_starts( &sha_ctx , 0 );
+
+        if( eddsa_id == MBEDTLS_EDDSA_CTX )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 0, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+        else if( eddsa_id == MBEDTLS_EDDSA_PREHASH )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 1, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+
+        /* tmp_buf contains the R point */
+        mbedtls_sha512_update( &sha_ctx, tmp_buf, sizeof( tmp_buf ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_point_write_binary( grp, Q, MBEDTLS_ECP_PF_COMPRESSED, &olen, tmp_buf, sizeof( tmp_buf ) ) );
+        mbedtls_sha512_update( &sha_ctx, tmp_buf, sizeof( tmp_buf ) );
+
+        /* In EDDSA_PREHASH, buf should contain the SHA512 hash. It contains the whole message otherwise */
+        mbedtls_sha512_update( &sha_ctx, buf, blen );
+
+        mbedtls_sha512_finish( &sha_ctx, sha_buf );
+        mbedtls_sha512_free( &sha_ctx );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary_le( &h, sha_buf, sizeof( sha_buf ) ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &h, &h, &grp->N ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_mul( grp, &sB, s, &grp->G, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_ecp_mul( grp, &hA, &h, Q, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_ecp_add( grp, &R, &R, &hA ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_point_write_binary( grp, &hA, MBEDTLS_ECP_PF_COMPRESSED, &olen, tmp_buf, sizeof( tmp_buf ) ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mpi( &sB.X, &sB.X, &R.X ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &sB.X, &sB.X, &grp->P ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mpi( &sB.Y, &sB.Y, &R.Y ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &sB.Y, &sB.Y, &grp->P ) );
+
+        if( mbedtls_mpi_cmp_int( &sB.X, 0) != 0 || mbedtls_mpi_cmp_int( &sB.Y, 0 ) != 0 )
+        {
+            ret = MBEDTLS_ERR_ECP_VERIFY_FAILED;
+            goto cleanup;
```

Comment:
Please use `mbedtls_ecp_point_cmp()` instead.

Reply:
After reading the paper again, I suggest to implement instead the "Fast single-signature verification" algorithm:

> To verify a single signature we use standard techniques for double-scalar multiplication to compute SB - H(R, A, M)A,
> and we then check whether the result is the same as R. (We actually check whether the encoding of the result is the same as the encoding of R, so that we can skip decompression of R.)

That replaces the call to `mbedtls_ecp_point_read_binary` by a call to `mbedtls_ecp_point_write_binary` which is way faster as it only contains some bit manipulation instead of arithmetics.